### DISCUSSION TITLE
How exactly does the runtime work?

### DISCUSSION THREAD

**Discussion by Grafcube** [slint-ui/slint#4558] [Q&A] [answered]
Specifically, I'm wondering how the native UI is done. I can see that the README shows different backends and styles and how to configure them and I also spent a few hours messing around with the examples.

My current understanding is that the UI markup is compiled beforehand as much as possible and the rest is controlled at runtime and the actual UI is left to native toolkits as much as possible.

1. What exactly is a `.slint` file compiled to for different platforms (linux, android etc.), backends (femtovg, skia) and styles (qt, cosmic, cupertino)? Since I got a single executable, I'm guessing that was bundled in and the QT build used my system's installation (since it used Breeze Dark).
2. Is it actually a native UI or just a theme? I'm guessing it's the latter because reimplementing the QT widgets sounds pointless and it only works if QT is installed to begin with. Basically, is it more similar to React Native than Flutter?
3. I was able to get Fluent, Cupertino and Material on my linux device so I'm wondering if it actually uses native libraries or bundles something for other platforms like Mac, Windows and Android. This is part of the reason I wasn't 100% sure if those were actually native widgets or just made to look like it.
4. Am I correct in my understanding that everything is modular i.e. the backend and style can be freely swapped. Is it also modular in rendering stack (like Vulkan instead of OpenGL or DirectX)?
5. What specifically does the runtime do (from the architecture diagram in the README)?
6. Minor question: Is making a DOM for native web apps theoretically possible? There are wasm based libraries like leptos and I was wondering if something like that was possible with Slint.

Thanks in advance. It's very likely that I'm just blind but I couldn't figure out where to look in the codebase to answer these questions myself. Sorry for the wall of questions.

**Comment by ogoffart** [comment, marked as answer]
1. A .slint file is compiled to either Rust or C++, it compiles into datastructure and code that calls the private API of the Slint runtime library.  Note that the style is itself composed of .slint file and compiled with it. (that's why style cannot be changed at runtime at the moment)  the Qt style will call into the Slint's qt backend private API which will use Qt's QStyle to draw the widgets.

2. "Just a theme".  (Apart for the Qt style that uses Qt itself)

3. It uses some native libraries for fonts loading or connecting to the display server (eg, x11), or Qt if using the Qt backend.

4. Correct. The renderer can be swapped at runtime (if compiled in.)  But at this point the Style cannot. Hopefully in the future we can change the style at runtime as well.

5. It traverse the item tree in order to render, it dispatches input event to the right data structure, it runs the event loop.

6. Yes, this is theoretically possible. Instead of rendering as open GL command, we would need to do a tree-diffing approach and generate DOM node.

Hope this helps.

**Reply by Grafcube** [reply]
Thanks! This clears up a lot of other questions I had as well. That just leaves me with these two questions.

1. I know that Android support is still a WIP but is it going to use native android widgets or do what Flutter does and render everything manually.
2. Where exactly can I find the code for the QT backend here? Is it theoretically possible to use the same method to make use of platform native UI libraries (like win32, cocoa, material you etc.)?

**Reply by ogoffart** [reply]
1. It will do what flutter does and render them 

2. https://github.com/slint-ui/slint/blob/master/internal/backends/qt/qt_widgets.rs and https://github.com/slint-ui/slint/blob/master/internal/backends/qt/qt_widgets  
 Yes it may be possible to use the platform code. We've considered doing that: https://github.com/slint-ui/slint/issues/43 and https://github.com/slint-ui/slint/issues/44
